core/src/main/java/org/infinispan/persistence/sifs/Index.java

```diff
@@ -245,46 +275,70 @@ public void start(Executor executor) {
       private final ReadWriteLock rootLock = new ReentrantReadWriteLock();
       private final boolean loaded;
       private final FileChannel indexFile;
+      private final File indexCountFile;
       private long indexFileSize;
-      private final AtomicLong size = new AtomicLong();
+      final AtomicLongArray sizePerSegment;
```

Comment:
> IMO this "clash" between SIFS segment and hash segment deserves a comment.

Yeah, unfortunately, the original SIFS called these as segments. It was a poor choice when it was originally added.

> Actually, do we really need one sizePerSegment for each SIFS-segment? Couldn't we maintain a single sizePerSegment for the entire store? Would it be cause problems if the index is ok for one segment but has to be rebuilt for another?

I need to think more about condensing the sizePerSegment, but it may work.

Reply:
I wasn't thinking of doing it for the memory, I didn't think it would be that much (especially now that the number of index segments is back to 3). OTOH I don't think having to look for each cache segment in each index segment would be a problem either, because the different hashing algorithms would make it almost impossible to have all the keys of a CH segment in the same index segment.

At the time I thought the sizes might even be maintained in the NBSIFS class so it would be clear what segments it's about, and so you wouldn't need the comment after all ;)

Reply:
In changing this it uncovered a separate bug related if the size changed concurrently with the initial check. Still fixing.